Auto-generate rule name (#1451)
vmware-tanzu/antrea: docs/antrea-network-policy.md

**ingress**: Each ClusterNetworkPolicy may consist of zero or more ordered
set of ingress rules. Each rule, depending on the `action` field of the rule,
allows or drops traffic which matches both the `from` and `ports` sections.

> Also, each rule has an optional `name` field, which should be unique within 

Reviewer: Question - why rule name must be unique? I think it does really matter much even two rules have the same name (thought some features use rule name to identify a rule might get incorrect results).
@abhiraut

Reviewer 2: i think this requirement came from @tnqn so we could use rule name to identify a rule to report stats per rule

Reviewer: I guess you first need a way to indicate a name is auto-generated?

Reviewer 3: Normally names of subresources of a resource are required to be unique within the resource. It helps status report for subresources and also makes the field "name" meaningful.
https://github.com/kubernetes/kubernetes/blob/096819c9635035abb4c4ec2bcaf4db8c35250cb7/pkg/apis/core/types.go#L48
https://github.com/kubernetes/kubernetes/blob/096819c9635035abb4c4ec2bcaf4db8c35250cb7/pkg/apis/core/types.go#L1752
https://github.com/kubernetes/kubernetes/blob/096819c9635035abb4c4ec2bcaf4db8c35250cb7/pkg/apis/core/types.go#L2095
https://github.com/kubernetes/kubernetes/blob/096819c9635035abb4c4ec2bcaf4db8c35250cb7/pkg/apis/core/types.go#L3707

**ingress**: Each ClusterNetworkPolicy may consist of zero or more ordered
set of ingress rules. Each rule, depending on the `action` field of the rule,
allows or drops traffic which matches both the `from` and `ports` sections.
Also, each rule has an optional `name` field, which should be unique within 
the policy describing the intention of this rule. If `name` is not provided for 
a rule, it will be auto-generated by Antrea. The auto-generated name will be 
of format `[ingress/egress]-[action]-[uid]`, e.g.

> ingress-allow-2f0ed6ec1949367d0107ea931043df9eb430b1c3,

Reviewer 3: could we use a shorter suffix? I think the chance of collision is small enough even we only use the first 6 chars (36^6=2176782336) given the name only needs to be unique within a policy and it has direction and action as prefix. (when generating network interface names for Pods, we do similar hash and use the first 6 chars as suffix). A short suffix looks cleaner and makes it possible to type it.

Author: Make sense. I can change it. 
BTW do we need an additional check for this collision or just assume it won't happen?

Reviewer 4: we should do 7, the default short git sha length :)

> the policy describing the intention of this rule. If users didn't provide `name`,
> antrea will auto-generate it. The auto-generate name will be in format: 
> ingress-[action]-[uid], e.g., 
> ingress-allow-2f0ed6ec1949367d0107ea931043df9eb430b1c3.
> The uid will be hash value of the rule based on sha1 algorithm. If a policy 

Reviewer 5: ```suggestion
the policy describing the intention of this rule. If `name` is not provided for 
a rule, it will be auto-generated by Antrea. The auto-generated name will be 
of format `ingress-[action]-[uid]`, e.g.
ingress-allow-2f0ed6ec1949367d0107ea931043df9eb430b1c3,
where [uid] is the hash value of the rule based on sha1 algorithm. If a policy 
```

> contains duplicate rules, or a rule use the auto-generated name of other rules 
> as its name, it will cause a collision. 

Reviewer 5: ```suggestion
contains duplicate rules, or if a rule name is same as the auto-generated name of
some other rule in the same policy, it will cause a conflict. 
```

> some other rules in the same policy, it will cause a conflict. 

Author: Thx.

**egress**: Each ClusterNetworkPolicy may consist of zero or more ordered set
of egress rules. Each rule, depending on the `action` field of the rule, allows
or drops traffic which matches both the `to` and `ports` sections. 
Also, each rule has an optional `name` field, which should be unique within 
the policy describing the intention of this rule. If users didn't provide `name`,

> antrea will auto-generate it. The auto-generate name will be in format: 

Reviewer 5: we could simply say something like 'the same rule name auto-generation process is applied to egress rules'
